### PR TITLE
Fix bottom sheet rebuilding when tapping

### LINKED ISSUE #126833: [3.10.0] ModalBottomSheet rebuilds on drag
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Execute the code sample.
2. Click on top of modalbottomsheet

### Expected results

The build function must work once

### Actual results

Rebuild on every clicking on ModalBottomSheets

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: MyApp()));

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: OutlinedButton(
          child: Text('openModalBottomSheet'),
          onPressed: () {
            showModalBottomSheet(
                context: context,
                builder: (ctx) {
                  print('rebuild');
                  return Text('data');
                });
          },
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>
<video  src="https://github.com/flutter/flutter/assets/52926902/d1eacc13-66f5-4d26-b2d0-0d4faa11d3d5.mov" controls>

</video>

</details>

### Logs

<details><summary>Logs</summary>

```console
D/EGL_emulation( 8150): app_time_stats: avg=20413.49ms min=20413.49ms max=20413.49ms count=1
2 I/flutter ( 8150): rebuild
D/EGL_emulation( 8150): app_time_stats: avg=40.72ms min=11.72ms max=653.52ms count=27
8 I/flutter ( 8150): rebuild
D/EGL_emulation( 8150): app_time_stats: avg=136.06ms min=81.13ms max=215.31ms count=8
3 I/flutter ( 8150): rebuild
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.0, on macOS 13.3 22E252 darwin-arm64, locale ru-RU)
    • Flutter version 3.10.0 on channel stable at /Users/jarvis/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 84a1e904f4 (6 days ago), 2023-05-09 07:41:44 -0700
    • Engine revision d44b5a94c9
    • Dart version 3.0.0
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/jarvis/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.77.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.64.0

[✓] Connected device (2 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554 • android-arm64 • Android 12 (API 31) (emulator)
    • macOS (desktop)             • macos         • darwin-arm64  • macOS 13.3 22E252 darwin-arm64

[✓] Network resources
    • All expected network resources are available.
```

</details>

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
This fixes an issue where the bottom sheet would rebuild when `enableDrag` is set to true on every tap. This is because `DragGestureRecognizer` would win the arena by default and dispatch the `drag` callbacks (in `acceptGesture`) even though it had not met the drag threshold. This changes keep the default behavior of `DragGestureRecognizer` the same, but adds a parameter `onlyAcceptDragOnThreshold` that a user can use to stop drag callbacks from being fired when the drag threshold has not been met.

Fixes #126833

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/gestures/monodrag.dart [hunk 1]
```diff
@@ -210,6 +211,18 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
...
   double? maxFlingVelocity;
 
+  /// Whether the drag callbacks should only be dispatched when the drag threshold
+  /// has been met.
+  ///
+  /// If true, the drag callbacks will only be dispatched when this recognizer has
+  /// won the arena and the drag threshold has been met.
+  ///
+  /// If false, the drag callbacks will be dispatched when this recognizer has won
+  /// the arena.
```
Comment (justinmc) [inline packages/flutter/lib/src/gestures/monodrag.dart:221]: Nit: Maybe say "...will be dispatched immediately when..."

At first glance this seems to say almost the same thing as the paragraph above.

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 2]
```diff
@@ -1300,6 +1298,39 @@ PersistentBottomSheetController<T> showBottomSheet<T>({
...
+              ..onUpdate = onVerticalDragUpdate
+              ..onEnd = onVerticalDragEnd
+              ..onlyDispatchDragCallbacksWhenThresholdMet = true;
+          },
+        ),
+      },
+      child: child,
+    );
+  }
+}
```
Comment (justinmc) [inline packages/flutter/lib/src/material/bottom_sheet.dart:1335]: Nit: Too many newlines here (maybe not your PR's fault).

File: packages/flutter/lib/src/gestures/monodrag.dart [hunk 3]
```diff
@@ -401,45 +420,8 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
...
-          primaryDelta: _getPrimaryValueFromOffset(localUpdateDelta),
-          globalPosition: correctedPosition.global,
-          localPosition: correctedPosition.local,
-        );
-      }
-      // This acceptGesture might have been called only for one pointer, instead
-      // of all pointers. Resolve all pointers to `accepted`. This won't cause
-      // infinite recursion because an accepted pointer won't be accepted again.
-      resolve(GestureDisposition.accepted);
+    if (onlyDispatchDragCallbacksWhenThresholdMet ? _hasDragThresholdBeenMet : !onlyDispatchDragCallbacksWhenThresholdMet) {
```
Comment (justinmc) [inline packages/flutter/lib/src/gestures/monodrag.dart:423]: Nit: Is this any easier to understand?

```suggestion
    if (onlyDispatchDragCallbacksWhenThresholdMet ? _hasDragThresholdBeenMet : true) {
```

File: packages/flutter/lib/src/gestures/monodrag.dart [hunk 4]
```diff
@@ -77,6 +77,7 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
     super.debugOwner,
     this.dragStartBehavior = DragStartBehavior.start,
     this.velocityTrackerBuilder = _defaultBuilder,
+    this.onlyDispatchDragCallbacksWhenThresholdMet = false,
```
Comment (justinmc) [inline packages/flutter/lib/src/gestures/monodrag.dart:80]: Just checking, is this the same default behavior as before?

File: packages/flutter/lib/src/gestures/monodrag.dart [hunk 5]
```diff
@@ -210,6 +211,18 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
...
+  /// has been met.
+  ///
+  /// If true, the drag callbacks will only be dispatched when this recognizer has
+  /// won the arena and the drag threshold has been met.
+  ///
+  /// If false, the drag callbacks will be dispatched when this recognizer has won
+  /// the arena.
+  ///
+  /// This value defaults to false.
+  bool onlyDispatchDragCallbacksWhenThresholdMet;
```
Comment (justinmc) [inline packages/flutter/lib/src/gestures/monodrag.dart:224]: Nit: What if this was something like `double threshold = gestureSettings?.touchSlop;`?

Or if that's not possible, maybe shorten the name to "useThreshold" or something?

I'm just throwing out ideas, feel free to ignore all of this :)

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1:

File: packages/flutter/lib/src/gestures/monodrag.dart (hunk 5 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/gestures/monodrag.dart:224]: I don't think this is possible. There is already a threshold being checked in `_hasSufficientGlobalDistanceToAccept` in `handleEvent`, the issue is that if the recognizer is in the arena by itself it gets accepted by default. In the case of bottom sheet this happens pretty much immediately on the first `PointerDownEvent`/ tap. `acceptGesture` is then called which fires the `drag*` callbacks. The threshold check never happens before acceptance. 

I do agree though this needs a better name.

maybe `onlyAcceptDragOnThreshold`?

File: packages/flutter/lib/src/gestures/monodrag.dart (hunk 4 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/gestures/monodrag.dart:80]: That is correct.

File: packages/flutter/lib/src/gestures/monodrag.dart (hunk 3 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/gestures/monodrag.dart]: Changed to `!onlyAcceptDragOnThreshold || _hasThresholdBeenMet`

File: packages/flutter/lib/src/gestures/monodrag.dart [hunk 6]
```diff
@@ -210,6 +211,17 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
   /// If null then [kMaxFlingVelocity] is used.
   double? maxFlingVelocity;
 
+  /// Whether the drag threshold should be met before dispatching any drag callbacks.
```
Comment (justinmc) [inline packages/flutter/lib/src/gestures/monodrag.dart:214]: Nit: Should you explain what "meeting the drag threshold" means?  Or maybe link to `touchSlop`?

**Review by justinmc — APPROVED** [review]
LGTM :+1:

File: packages/flutter/lib/src/gestures/monodrag.dart (hunk 6 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/gestures/monodrag.dart:214]: Done. I linked `computeHitSlop` and `computePanSlop` as they determine the threshold.

**Pull request merged**